### PR TITLE
Add EvolvedAttention: A transformer-based neural network strategy for Prisoner's Dilemma

### PR DESCRIPTION
**Pull request by moderouin**
## Description
This PR introduces the `EvolvedAttention` strategy, a novel approach that uses a transformer neural network with self-attention mechanisms to make decisions. The strategy analyzes game history through attention patterns to determine optimal moves.

## Features
- Transformer architecture with self-attention (24 layers, 8 attention heads)
- Memory depth of 200 moves with reverse-chronological processing
- GPU acceleration when available (with CPU fallback)
- Pre-trained model weights from evolutionary self-play

## Technical Implementation
- Game states encoded as tokens (CC, CD, DC, DD plus special CLS/PAD tokens)
- Position embeddings to maintain sequence information
- Decision boundary using sigmoid activation (< 0.5 → Cooperate, ≥ 0.5 → Defect)

## Performance Considerations
The neural network is relatively complex but runs efficiently on modern hardware. The strategy balances analytical depth with reasonable computational requirements.

### CODE REVIEW

**Comment by marcharper** [comment]
Hi @moderouin, thanks for your contribution! Have you run the tests locally? I'm wondering if the new tests are very slow or if the issue is with Github's CI.

**Comment by moderouin** [comment]
The test runs within 20 minutes on my PC with Ubuntu 22.04.5 LTS and also seems to work well on Windows and macOS. However, I’m able to reproduce the CI bug within a Docker container using `ubuntu:latest`; the tests never seem to end. It appears to be an issue with `*.rts`. I’m currently trying to fix this, but if you have any insights, that would be great! @marcharper

**Comment by moderouin** [comment]
It seems like the error was caused by the forking of process on Linux. I use spawn instead to avoid any deadlocks @marcharper

**Comment by marcharper** [comment]
How does the strategy perform? Have you run any tournaments?

**Comment by moderouin** [comment]
When running a tournament with 10 repetitions against all the strategies not in long_run_time, it ranked first in median score, with a score close to that of EvolvedLookerUp2_2_2.

**Comment by marcharper** [comment]
Can you tell us more about how you trained it?

**Comment by moderouin** [comment]
This strategy was trained by performing multiple rounds of tournaments against all strategies and the current network of EvolvedAttention. After each tournament, the strategy learns to reproduce the moves of the best-performing strategy from the last tournament.

**Comment by moderouin** [comment]
I’m also currently working on a second version with the same architecture, but incorporating an actor-critic approach with policy and value heads on top of the base network and training it using PPO.

**Comment by moderouin** [comment]
Is there any adjustment I should make to this strategy? @marcharper

**Comment by marcharper** [comment]
No, I just haven't had a chance to review thoroughly, and we require two maintainer review. It's not surprising that it does so well if you trained against all the other strategies (it might be overfit), but that's not a blocker to including it. For your future trainings, try using just the short runtime strategies, which has worked fine for the other ML strategies and saves a lot of computation time.

File: axelrod/data/all_classifiers.yml [hunk 1]
```diff
@@ -439,6 +439,14 @@ Evolved ANN 5 Noise 05:
   manipulates_state: false
   memory_depth: .inf
   stochastic: false
+EvolvedAttention:
+  inspects_source: false
+  long_run_time: false
```
Comment (marcharper) [inline axelrod/data/all_classifiers.yml]: Probably should be true. We don't have a formal definition of long run time, but based on the size of the parameters I'm guessing this one is relatively slow compared to most of the strategies in the library.

File: axelrod/strategies/attention.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+
+    Names:
+    - EvolvedAttention: EvolvedAttention by Marc-Olivier Derouin
+    """
+
+    name = "EvolvedAttention"
+    classifier = {
+        "memory_depth": MEMORY_LENGTH,
+        "stochastic": False,
+        "long_run_time": False,
```
Comment (marcharper) [inline axelrod/strategies/attention.py]: Same here

File: axelrod/strategies/attention.py [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+    player_history = player.history[-MEMORY_LENGTH:]
+    player_history = player_history[::-1]
+    opponent_history = opponent.history[-MEMORY_LENGTH:]
+    opponent_history = opponent_history[::-1]
+
+    feature_size = MEMORY_LENGTH + 1 if right_pad else len(player_history) + 1
+
+    game_history = torch.full((feature_size,), PAD_TOKEN, dtype=torch.int)
+    game_history[0] = CLS_TOKEN
+    for index, (action_player, action_opponent) in enumerate(
```
Comment (marcharper) [inline axelrod/strategies/attention.py:54]: This seems inefficient, since it's called every round and loops through the entire history, so it's going to be repeating a lot of calculations, roughly O(N^2) in the length of the match. IIUC you could cache the calculations (in the Player class) and then just append the latest calculation each time to make this linear.

File: axelrod/tests/integration/test_matches.py [hunk 4]
```diff
@@ -29,7 +29,7 @@ class TestMatchOutcomes(unittest.TestCase):
         ),
         turns=integers(min_value=1, max_value=20),
     )
-    @settings(max_examples=5)
+    @settings(max_examples=5, deadline=None)
```
Comment (marcharper) [inline axelrod/tests/integration/test_matches.py]: If these tests are slow perhaps it's better to lower max examples. @drvinceknight wdyt?

File: axelrod/tournament.py [hunk 5]
```diff
@@ -3,7 +3,9 @@
 import os
 import warnings
 from collections import defaultdict
-from multiprocessing import Process, Queue, cpu_count
+from multiprocessing import Process, Queue, cpu_count, set_start_method
+
+set_start_method("spawn", force=True)
```
Comment (marcharper) [inline axelrod/tournament.py:10]: Can you add a comment why this is needed?

File: tox.ini [hunk 6]
```diff
@@ -28,6 +28,7 @@ deps =
     isort
     black
     numpy==1.26.4
```
Comment (marcharper) [inline tox.ini:30]: @drvinceknight is there a reason we've fixed the versions here?

**Review by marcharper — CHANGES_REQUESTED** [review]
Looks good overall, some minor comments. PTAL and thanks for the contribution!

**Comment by marcharper** [comment]
@jsafyan PTAL, you're more of a transformer expert than I.

File: axelrod/strategies/attention.py (hunk 3 above)
Comment (moderouin) [reply axelrod/strategies/attention.py:54]: Yes, this is ineffective, but I wasn’t sure if we wanted to store the features in the player class. This means we will also need to override the implementation of update_history, since we need to update the features tensor.

Is that what we want?

Considering that the processing time of compute_features is not the bottleneck of this strategy and that there are other operations with O(N²) complexity, this seems acceptable.

Comment (marcharper) [reply axelrod/strategies/attention.py:54]: It might be a bottleneck, if someone tries to run a tournament with a large match length, so if there's a linear alternative that's preferable. Or is there something else here quadratic in the match length?

It's fine to store something locally, many strategies do that. I don't think you'd have to overwrite `update_history`, just update the features tensor in the strategy function at the start of each round.

Comment (moderouin) [reply axelrod/strategies/attention.py:54]: After looking further into the repo, I understand why I don't need to overwrite update_history.

So even though someone might want to test a large match length, this strategy still checks the last 200 steps. As a result, the runtime will remain the same, even if we run a tournament with a large match length.

I need to insert the last game state at index 1 within the tensor, so I have to shift every element within [1:]. This means that even though I store the previous features, I still have an O(n) operation at each round.

Anyway, the overall strategy has a complexity of roughly O(n^3) in the length of the match, since there are n attention computations, each of which is O(n^2) (attention mechanism).

Comment (marcharper) [reply axelrod/strategies/attention.py:54]: Ok, let's mark the strategy as `long_run_time` (see the other comments) and leave this as is.

**Comment by moderouin** [comment]
@marcharper Do you have an idea why this test fail in the last check? Seems like it's not related to my strategy?

File: tox.ini (hunk 6 above)
Comment (drvinceknight) [reply tox.ini:30]: It was done here to avoid issues with 2.0: https://github.com/Axelrod-Python/Axelrod/pull/1446

File: axelrod/tests/integration/test_matches.py (hunk 4 above)
Comment (drvinceknight) [reply axelrod/tests/integration/test_matches.py]: Yeah, even if we go down to 2 that's not a bad idea.

**Comment by moderouin** [comment]
Is there anything else I should change ? @drvinceknight @marcharper

**Comment by drvinceknight** [comment]
> Is there anything else I should change ?

Would you be able to add a few more tests (against known sequences of play or strategies) here axelrod/tests/strategies/test_attention.py please?

I see you have the test again two cooperations but it would be good to have more wide ranging ones (even if due to the nature of this strategy it serves more as a regression test).

**Comment by moderouin** [comment]
Let me know what you think @drvinceknight

**Review by drvinceknight — APPROVED** [review]
This looks good to me.

**Comment by drvinceknight** [comment]
@marcharper I think this is just waiting on you to ok.

**Comment by marcharper** [comment]
The entire test suite is substantially slower now. Is there something we can do to mitigate this, such as loading the weights more lazily?

**Comment by drvinceknight** [comment]
> The entire test suite is substantially slower now. Is there something we can do to mitigate this, such as loading the weights more lazily?

**Comment by drvinceknight** [comment]
(Sorry I clicked a button by mistake closing the PR!)

**Comment by drvinceknight** [comment]
> The entire test suite is substantially slower now. Is there something we can do to mitigate this, such as loading the weights more lazily?

I hadn't realised this. How mush slower is it? If @marcharper's good suggestion doesn't work what about marking some tests as "slow" (pytest has tooling for this: https://pytest-with-eric.com/pytest-best-practices/pytest-markers/)?

**Comment by moderouin** [comment]
> The entire test suite is substantially slower now. Is there something we can do to mitigate this, such as loading the weights more lazily?

Looking into this

**Comment by moderouin** [comment]
Lazy loading of the model weights has reduced the total test run time by roughly one hour. Let me know what you think @marcharper

**Comment by drvinceknight** [comment]
Unless I'm not remembering correctly before this PR the test run would be about 20 minutes, and your latest commit is at 60 minutes.

Have I got that right? (not sure who this question is for)

**Comment by moderouin** [comment]
> Unless I'm not remembering correctly before this PR the test run would be about 20 minutes, and your latest commit is at 60 minutes.
> 
> Have I got that right? (not sure who this question is for)

That's correct. `test_meta.py` is taking a really long time to run. I'll try to reduce the runtime while still preserving the quality of the tests

**Comment by marcharper** [comment]
@drvinceknight I'm ok with this going in for now, we should think about how to speed up the test suite though, maybe by ensuring that fewer of the long run time strategies end up in test tournaments or meta strategies.

**Comment by drvinceknight** [comment]
> That's correct. `test_meta.py` is taking a really long time to run. I'll try to reduce the runtime while still preserving the quality of the tests

Thanks I think you've done an awesome job, appreciate you working to get the tests under control :)

> @drvinceknight I'm ok with this going in for now, we should think about how to speed up the test suite though, maybe by ensuring that fewer of the long run time strategies end up in test tournaments or meta strategies.

100%!
